toolchain/codegen/codegen.h

```diff
@@ -8,15 +8,44 @@
 #include <cstdint>
 
 #include "llvm/IR/Module.h"
+#include "llvm/Target/TargetMachine.h"
 
 namespace Carbon {
-// Prints the assembly to stdout for the given llvm module.
-// If print fails, this returns false and any information about the failure is
-// printed to the error stream.
-auto PrintAssemblyFromModule(llvm::Module& module,
-                             llvm::StringRef target_triple,
-                             llvm::raw_pwrite_stream& error_stream,
-                             llvm::raw_pwrite_stream& output_stream) -> bool;
+class CodeGen {
+ public:
+  CodeGen(llvm::Module& module, llvm::StringRef triple,
+          llvm::raw_pwrite_stream& error_stream,
+          llvm::raw_pwrite_stream& output_stream)
+      : Module(module),
+        output_stream_(output_stream),
+        error_stream_(error_stream),
+        target_triple(triple){};
+  // Generates the object code file.
+  // Returns false in case of failure, and any information about the failure is
+  // printed to the error stream.
+  auto GenerateObjectCode(llvm::StringRef output_file) -> bool;
+
+  // Prints the assembly to stdout.
+  // Returns false in case of failure, and any information about the failure is
+  // printed to the error stream.
+  auto PrintAssembly() -> bool;
+
+ private:
+  llvm::Module& Module;
+  llvm::raw_pwrite_stream& output_stream_;
+  llvm::raw_pwrite_stream& error_stream_;
+  llvm::StringRef target_triple;
+  // Creates the target machine for triple.
+  // Returns nullptr in case of failure, and any information about the failure
+  // is printed to the error stream.
+  auto CreateTargetMachine() -> llvm::TargetMachine*;
+  // Using the llvm pass emits either assembly or object code to dest.
+  // Returns false in case of failure, and any information about the failure is
+  // printed to the error stream.
+  auto EmitCode(llvm::raw_pwrite_stream& dest,
+                llvm::TargetMachine* target_machine,
+                llvm::CodeGenFileType file_type) -> bool;
```

Comment:
```suggestion
  // Creates the target machine for triple.
  // Returns nullptr in case of failure, and any information about the failure
  // is printed to the error stream.
  auto CreateTargetMachine() -> llvm::TargetMachine*;

  // Using the llvm pass emits either assembly or object code to dest.
  // Returns false in case of failure, and any information about the failure is
  // printed to the error stream.
  auto EmitCode(llvm::raw_pwrite_stream& dest,
                llvm::TargetMachine* target_machine,
                llvm::CodeGenFileType file_type) -> bool;

  llvm::Module& Module;
  llvm::raw_pwrite_stream& output_stream_;
  llvm::raw_pwrite_stream& error_stream_;
  llvm::StringRef target_triple;
```